### ISSUE TITLE
Cannot rewrite valid HIERARCH keyword.

### ISSUE THREAD

**Issue by vforchi** [nom-tam-fits/nom-tam-fits#695]
One of the ESO pipelines creates a file with this keyword:
`HIERARCH ESO PRO REC2 CAL2 NAME= 'ESPRESSO_SINGLEHR_2x1_reduced_inst_config_201'`
which is exactly 80 characters long. The file has been created with cfitsio.

When I open it with nom.tam.fits, modify some other keyword and try to write it, I get an error because the formatter adds a space after the keyword name.

According to what I found in the code of cfitsio and fitsverify, that keyword is valid.

**Comment by attipaci** [comment]
Hi @vforchi. It's a feature, not a bug...

You are correct that it can be a valid 80-character header entry. The extra space before the `=` was added a long-long time ago (before my time) specifically based on a user's request, who could not read with CFITSIO the files produced by nom-tam-fits otherwise. Back then CFITSIO expected a space before the `=` and barfed if it was not there. The HIERARCH convention certainly does not forbid the extra space, so Tom or Ritchie have added it to conform with CFITSIO back then. 

It looks like CFITSIO has changed how it handles HIERARCH keywords since then. That's great! In principle now, we can change nom.tam.fits to no longer add the extra space, but that might break usage with older CFITSIO tools. So, it would be preferable not to go down that route.

In this particular case, however, given that the entry is a string, it should be possible to write it with the long string convention (which is enabled by default). The error can thus result from two reasons:

1. Perhaps the ESO pipeline explicitly disables that feature with `FitsFactory.setLongStringsEnabled(false)`, or else
2. the __nom.tam.fits__ library has a bug in not realizing that the above HIERARCH keyword should be stored using the long string convention.

If it's #1, the obvious fix would be for the ESO pipeline to allow the long string convention (which is part of the current, and by now several years old, FITS standard). And if it's #2, then it's a bug which I can (and will) fix promptly...

There is a 3rd route also, which is for the application to define it's own way of formatting HIERARCH keywords. You can simply implement the `nom.tam.fits.header.hierarch.IHierarchKeyFormatter` interface (or else extend the `.StandardIHierarchKeyFormatter` class) to write HIERARCH keywords just the way you want/need. Then you can call `FitsFactory.setHierarchFormatter()` to activate it. 

I'd be happy to provide such a new formatter class (e.g. named `ESOHierarchKeyFormatter`) if that helps -- but I'm hesitant to make it the default one for the mentioned reasons that it might break existing usage. So, the ESO pipeline would still have to call `FitsFactory.setHierarchFormatter()` probably to activate it. Let me know if that is your preferred solution.

I hope this helps both to explain and to find a suitable solution. Let me know if you think the issue is #2, and I should provide a fix for that.

**Comment by attipaci** [comment]
PS. In the meantime, I'll contemplate the impact of removing the extra space from the default HIERARCH formatter in future releases. Maybe given that CFITSIO no longer expects it (hopefully for a long time now), perhaps it's pointless to worry about it, and so perhaps it's OK to change the default formating to that without the extra space...

**Comment by vforchi** [comment]
For me what is confusing that if you open a file and write it to another location, then the contents are different.
The pipeline uses cfitsio, which adds the space by default, but it does not if the keyword + value would fits in 80 char without it.

**Comment by vforchi** [comment]
Also, we cannot enable long strings because some of our tools do not support it, so we make sure that keyword name plus value always fits in 80 characters.

**Comment by attipaci** [comment]
I

> For me what is confusing that if you open a file and write it to another location, then the contents are different. The pipeline uses cfitsio, which adds the space by default, but it does not if the keyword + value would fits in 80 char without it.

Well, it comes from the fact that when you open the file you process the information from the header, and then when you re-write it (possibly after modifying some header values) then it is not just copied from the input, but rather it is recreated from the current set of key/value pairs. Both nom.tam.fits and CFITSIO (hopefully) follow the standard, but there is room for different intepretations of it, especially on where and how many empty spaces are used. So some differences are to be expected. You should expect exact reproduction only if you use the same tools repeatedly (self consistency). Blame the FITS standard for not being more specific if you will...

I'll see if I can make the HIERARCH formatter to behave the same way as CFITSIO. However, I do find it to be a bit of a madness that a space is added, or not, depending on the length of the value that follows it. It's CFITSIO that is not consistent. Why don't they just skip the space always? Yes, we can try to confirm to its whims, but it's hard when it's not documented anywhere and subject to change without notice...

**Comment by attipaci** [comment]
On a side note CFITSIO has way too many of such idiosyncracies, funky implementations, and outright bugs, which are not spelled out in the FITS standard. So on a philosophical level, if one is to implement a FITS library (such as nom.tam.fits), is is expected to follow the published standard (which nom.tam.fits does more so than CFITSIO), or is it expected to recreate all the nonsensical and buggy implementation of CFITSIO?

**Comment by vforchi** [comment]
I totally agree with you, and I'm not asking to re-implement all the idiosyncracies, but on the other hand I would like to:
- open an existing file and save it (which does not work with that keyword)
- ideally the saved file is identical to the original file.
One possible approach would be to store the original representation in the `HeaderCard`, and use that for serialization, modifying only if the card is updated: what do you think? If you like the approach I could make a PR.

**Comment by attipaci** [comment]
Hi @vforchi,

What you suggest may work, but I feel it's a can of worms. It's sounds easy to save the original 80-character representation of a header card when read from the stream to use it for write-back later, but then you have to also worry about:

- Header cards that use the CONTINUE long keywords convention, with multiple 80-character records.
- How do you detect changes to the card's key, value, or comment fields, or its type, after the read. If the user changes any of these, then you should not use the original card image, but create it from the current properties. I cannot think of an elegant foolproof way to detect changes post-reading in all circumstances.
- What do you do with possibly improper cards, which violate the FITS standard. Are you going to ignore the fact that they are bad, and just write them as such anyway?

I personally like the fact that nom.tam.fits actually fixes up not quite standard headers. Its header parser is more flexible than that of most tools (e.g. CFITSIO), and can process a lot of 'bad' headers. It's a nice feature that on re-writing 'bad' headers it actually makes them proper.

And while 'ideally the saved file is identical to the original file' sounds nice at first, I'd say that the library producing consistent output is perhaps just as important if not more so...  Not to mention, that the saved file is never really going to be identical to the original, since you would not use a tool to read and save a file, unless you modified it. The moment you make any changes, you are no longer going to be identical to the original. Also, FITS headers are were never designed to be some fixed set of byte sequence. They simply represent a set of keys and associated values, and all headers that provide the same set of key/value pairs are identical from the FITS point of view, regardless of the exact representation in the header space. It is totally OK to not write back header cards in the exact form they appear in the original.

Having said all that (and having slept on it), I'd suggest to try something simpler first: modify the standard HIERARCH formatter to not add an extra space before `=`, and see if your pipeline and CFITSIO are happy with that. (I'll prepare a PR you can test today...)

If not, I'd suggest that perhaps it's time to update the ESO toolkit to support the long strings, which have been part of the standard since 2018. (Note that HIERARCH is still not part of the standard per se -- so it's funny that your tools support something non-standard, while they cannot handle the standard properly). The point is that if you can use long strings all of the headache goes away, including that you no longer have to carefully choose string values, such as `ESPRESSO_SINGLEHR_2x1_reduced_inst_config_201` to fit into the fixed space provided by some specific tool (CFITSIO) anyway...

And if CFITSIO has issues with not having the space before `=`, then perhaps CFITSIO should be fixed to be more flexible, especially given its own inconsistent approach for adding that space sometimes, and not at other times...

**Comment by attipaci** [comment]
@vforchi,

See if PR #696 fixes the problem for you. If so, I think that the simplest, most standard compliant, and potentially least problematic solution.

**Comment by attipaci** [comment]
BTW, the [ESO HIERARCH convention](https://fits.gsfc.nasa.gov/registry/hierarch/hierarch.pdf) does not say anything about spaces around the `=` (before or after), but all examples shown in the document have spaces both before and after. The letter of the convention simply requires an '=' in the free format field to separate the keyword and value components.

It might be a good idea to prod the authors of the ESO specification to update the specification to be more explicit on whether the surrounding spaces are required or optional so as to ensure that all tools are on the same page on the requirements... 

Ideally we should receive clarification from them before merging PR #696, and depending on what their take is on the convention, you might want to notify CFITSIO's maintainers and/or modify the tools you use to comply accordingly...

Without clarity on the use of spaces around the '=' anything we do here might just make a bigger mess of things...

**Comment by vforchi** [comment]
My problem with that PR is that it would modify all the HIERARCH headers as written by the pipelines and the instrument software, and I would rather change the file as little as possible.

My understanding of the convention is that the spaces are optional both before and after the equal sign, but I agree that it is vague.

Adopting the long string convention is not realistic, because we have ~15 active instruments and ~15 pipelines that create FITS files, and they won't be upgraded for this problem.

I understand your point of view, and I agree with most of it, but if I can't preserve the file as much as possible I'm afraid I will stick to use nom.tam.fits only for read only purposes.

**Comment by attipaci** [comment]
Hi @vforchi,

I understand your situation all too well. I have been on that side many times. But as a maintainer of the library I have to think of what works for all users in general, rather than to cater to individual use cases with their own very specific quirks. 

> My problem with that PR is that it would modify all the HIERARCH headers as written by the pipelines and the instrument software, and I would rather change the file as little as possible.

Unfortunately, the only way you can ensure that is to use the same library from creation to end, through all modification steps. CFITSIO has its peculiarities, and so all it's quirks will stay intact as long as you use it at every step.

I'm also of the opinion that you should not be so worried about changing more than the minimal amount as files are modified. As long as they remain human readable and can be used the same by other tools, it should not matter. It's an idealist stance, more than a practical one... :-)

**Comment by vforchi** [comment]
This is what I was doing before, but it meant writing a small C program just to write 3 keywords and invoke it from a java program.
When I originally wrote the software, the java library was much less mature than it is now, so I was hoping to get rid of this piece of software, but I think I can't :(

**Comment by attipaci** [comment]
@vforchi, I have not completely given up on implementing something similar to CFITSIO for 'compressible' spacing around the `=`. I'll see if I can coerce PR #696 to do something similar... I'll let you known if I have something you can try...

**Comment by attipaci** [comment]
@vforchi, I had another idea. I guess what you want is a surgical way to edit headers. How about a `HeaderCard.rewrite()` function, which only rewrites that one card to the disk, without touching the rest of the header? So for example,

```java
  // Read FITS from file...
  Fits fits = new Fits("mydata.fits");
  Header h = fits.getHDU(0).getHeader();
  HeaderCard card = h.getCard("OBJECT");

  // change a specific card 
  card.setValue("Modified object");

  // Write back just the modified card to the file, without touching anything else.
  card.rewrite();
```

 (Of course, I'll have to be careful about handling long-string cards this way, but that should be doable...)

**Comment by vforchi** [comment]
Not really, what I need to do is to add 3 keywords and update the checksum. The keyword that failed was completely unrelated.

**Comment by attipaci** [comment]
I see. That does make it complicated as header additions directly to the file could conflict with changes made to it in the digested form before write-back. The checksumming too is based on writing the whole header (in its current form) out first to calculate the current sum (but it's at least possible to calculate sums on the file directly).  I'll think more to see if there is a foolproof way these can be done on the file directly, but those seem to be in conflict with the higher-level functionality of the library. You might be stuck with the C code for such low-level edits...

**Comment by attipaci** [comment]
@vforchi, I merged PR #696, which adds a flexible treatment of spaces around `=` for HIERARCH cards, sort of like CIFITSIO. This at least should 'fix' the original 'issue' that started the discussion, so your longish HIERARCH keyword should no longer throw an exception even if long strings are enabled.

**Comment by vforchi** [comment]
That's great news, thanks, I will give it a try.

**Comment by attipaci** [comment]
Ooops, I think there was an omission from PR #696, which should be fixed by PR #697...
